### PR TITLE
(dev/drupal#8) Migrate bower.json into composer.json via composer-downloads-plugin

### PR DESCRIPTION
**Pull request by totten**
Before
------

* JS and CSS libraries are downloaded via `bower`.
* `bower` is deprecated.
* `bower` is slow.
* `bower` requires installing NodeJS.
* `bower` requires installing Bower.

After
-----

* `composer install` downloads the `bower_components` folder.
* `composer` has a better cache.
* `composer` is fast.
* You can setup a dev build with fewer dependencies.
* All of the `bower_components` will land in the same canonical location, regardless of project structure and environment.

Comments
--------

~~This commit is flagged work-in-progress because it depends on an unreleased branch of a composer plugin. I just want to push it up for the moment so some people can see it. I've opened an issue https://github.com/LastCallMedia/ComposerExtraFiles/issues/2 to discuss the path to publication.~~ 

This depends on [civicrm/composer-downloads-plugin](https://github.com/civicrm/composer-downloads-plugin), which is an updated variant of [lastcall/composer-extra-files](https://github.com/LastCallMedia/ComposerExtraFiles/).  It has a few key characteristics:

* It does not require any root-level package configuration or third-party tools/runtimes.
* It produces a file-structure that closely matches the current one.
* It produces the same file-structure whether one (a) uses civicrm-core as root-package or (b) uses civicrm-core as a dependency inside a bigger project.
* It is fairly transparent and tweakable - anyone who wants to submit a PR to add/upgrade/remove JS/CSS dependencies should be able to do so.
* When we roll out [civicrm extension/composer bridge](https://lab.civicrm.org/infra/comex), extensions should be able to use it while being compatible with several different deployment scenarios.

Technical Details
---------------------

To try this out, I renamed my old `bower_components`, ran `composer`, and compared the new/old trees. The substance all looks good -- there are no *missing* or *differing* files. ~~However, there is some new detritus.~~  (*Edit: This has been resolved by adding support for `ignore` directives. The subsequent discussion has been revised to match.*)

```
$ mv bower_components{,.orig}
$ composer install
...
$ colordiff -ru bower_components{.orig,} -x .bower.json -x .composer-downloads -x .setupsh.ts

Only in bower_components.orig/checklist-model: test
Only in bower_components.orig/dc-2.1.x: bower.json
Only in bower_components.orig/font-awesome: bower.json
Only in bower_components.orig/jstree: bower.json
```

These appear to be trivial differences:

* For `checklist-model/test`, the old tree simply has an empty folder. There's no real content in either old or new tree.
* For the `bower.json` files, the new tree is more consistent with the `ignore` specification. The difference arises because (in the old tree) `bower` doesn't let you ignore `bower.json`. But we don't really need the file at runtime.

See also: https://lab.civicrm.org/dev/drupal/issues/8

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -85,7 +86,145 @@
       "bash tools/scripts/composer/phpword-jquery.sh"
     ]
   },
+  "repositories": {
+    "extrafiles": {
+      "type": "vcs",
+      "url": "https://github.com/totten/ComposerExtraFiles.git"
+    }
```
Comment (totten) [inline composer.json]: This part is temporary (for demonstration purposes). It needs to be squashed out once we have a more permanent publication of the revised plugin.

**Comment by totten** [comment]
I've pushed an update which imports a bunch of the `ignore` clauses from bower -- the updated description how shows only a couple trivial differences between the old+new trees.

Additionally, to improve readability, I consolidated all the `path` statements into default `path` formula.

File: composer.json [hunk 2]
```diff
@@ -85,7 +86,130 @@
...
+      },
+      "phantomjs-polyfill": {
+        "url": "https://github.com/conversocial/phantomjs-polyfill/archive/v0.0.2.zip"
+      },
+      "qunit": {
+        "url": "https://github.com/jquery/qunit/archive/v1.10.0.zip"
+      },
+      "select2": {
+        "comment": "This is pegged to a branch, which may have caching+update issues. Should switch to a tag.",
+        "url": "https://github.com/colemanw/select2/archive/stable/3.5.zip"
```
Comment (totten) [inline composer.json]: @colemanw Would you mind making a tag in the `colemanw/select2` repo? Maybe something like

```
git tag v3.5-civicrm-1.0 90d4b9ee388b33e17c9f58a2866a4765f00b9a55
```

**Comment by colemanw** [comment]
@totten I have rebased the latest upstream commits with my select2 repo and tagged it `v3.5-civicrm-1.0` per your suggestion.

**Comment by totten** [comment]
Cheers @colemanw. I've pushed an update to use the new tag.

File: composer.json (hunk 1 above)
Comment (totten) [reply composer.json]: @seamuslee001 I've updated this to use a published plugin, so it no longer needs the `repositories: ... type: vcs ...` clause.

**Pull request merged**